File: solutions/python3/0002.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy = ListNode(None)
        curr = dummy
        carry = 0

        while carry or l1 or l2:
            carry += (l1.val if l1 else 0) + (l2.val if l2 else 0)
            curr.next = ListNode(carry % 10)
            curr = curr.next
            carry //= 10
            if l1:
                l1 = l1.next
            if l2:
                l2 = l2.next

        return dummy.next

File: solutions/cpp/0002.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* dummy = new ListNode(NULL);
        ListNode* curr = dummy;
        int carry = 0;

        while (carry || l1 || l2) {
            carry += (l1 ? l1->val : 0) + (l2 ? l2->val : 0);
            curr->next = new ListNode(carry % 10);
            curr = curr->next;
            carry /= 10;
            if (l1) l1 = l1->next;
            if (l2) l2 = l2->next;
        }

        return dummy->next;
    }
};